src/workerd/io/trace-stream.c++

```diff
@@ -0,0 +1,821 @@
+#include <workerd/api/global-scope.h>
+#include <workerd/io/io-context.h>
+#include <workerd/io/trace-stream.h>
+#include <workerd/io/worker-interface.h>
+#include <workerd/jsg/jsg.h>
+#include <workerd/util/completion-membrane.h>
+#include <workerd/util/uuid.h>
+
+#include <capnp/membrane.h>
+
+namespace workerd::tracing {
+namespace {
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::Attribute& attribute) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("attribute"_kj));
+  obj.set(js, "name"_kj, js.str(attribute.name));
+
+  auto ToJs = [](jsg::Lock& js, const tracing::Attribute::Value& value) -> jsg::JsValue {
+    KJ_SWITCH_ONEOF(value) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        return js.str(str);
+      }
+      KJ_CASE_ONEOF(b, bool) {
+        return js.boolean(b);
+      }
+      KJ_CASE_ONEOF(d, double) {
+        return js.num(d);
+      }
+      KJ_CASE_ONEOF(i, int32_t) {
+        return js.num(i);
+      }
+    }
+    KJ_UNREACHABLE;
+  };
+
+  if (attribute.value.size() == 1) {
+    obj.set(js, "value"_kj, ToJs(js, attribute.value[0]));
+  } else {
+    auto values = KJ_MAP(val, attribute.value) { return ToJs(js, val); };
+    obj.set(js, "value"_kj, js.arr(values));
+  }
+
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, kj::ArrayPtr<const tracing::Attribute> attributes) {
+  auto attrs = KJ_MAP(attr, attributes) { return ToJs(js, attr); };
+  return js.arr(attrs);
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::FetchResponseInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("fetch"_kj));
+  obj.set(js, "statusCode"_kj, js.num(info.statusCode));
+  return obj;
+}
+
+jsg::JsValue ToJs(jsg::Lock& js, const tracing::FetchEventInfo& info) {
+  auto obj = js.obj();
+  obj.set(js, "type"_kj, js.str("fetch"_kj));
+  obj.set(js, "method"_kj, js.str(kj::str(info.method)));
+  obj.set(js, "url"_kj, js.str(info.url));
+  obj.set(js, "cfJson"_kj, js.str(info.cfJson));
```

Comment:
Q: Is there a JSG API that allows creating an object with an array of object keys and values?

Reply:
Not currently. Can look at adding that later.